File: src/db-toolkit/core/routes/query.py
"""Query execution routes."""

import json
from fastapi import APIRouter, HTTPException
from core.storage import ConnectionStorage
from core.schemas import QueryRequest, QueryResponse
from operations.query_executor import QueryExecutor
from operations.query_history import QueryHistory
from operations.explain_analyzer import ExplainAnalyzer

router = APIRouter()
storage = ConnectionStorage()
executor = QueryExecutor()
history = QueryHistory()
analyzer = ExplainAnalyzer()


@router.post("/connections/{connection_id}/query", response_model=QueryResponse)
async def execute_query(connection_id: str, request: QueryRequest):
    """Execute query on connection."""
    from operations.operation_lock import operation_lock

    connection = await storage.get_connection(connection_id)
    if not connection:
        raise HTTPException(status_code=404, detail="Connection not found")

    if operation_lock.is_locked(connection_id):
        raise HTTPException(
            status_code=409, detail="Connection is busy with another operation"
        )

    lock = operation_lock.get_lock(connection_id)
    async with lock:
        result = await executor.execute_query(
            connection=connection,
            query=request.query,
            limit=request.limit,
            offset=request.offset,
            timeout=request.timeout,
        )
    
        # Save to history
        history.add_query(
            connection_id=connection_id,
            query=request.query,
            success=result["success"],
            execution_time=result["execution_time"],
            row_count=result["total_rows"],
            error=result.get("error"),
        )

        return QueryResponse(**result)


@router.get("/connections/{connection_id}/query/history")
async def get_query_history(connection_id: str, limit: int = 50):
    """Get query history for connection."""
    connection = await storage.get_connection(connection_id)
    if not connection:
        raise HTTPException(status_code=404, detail="Connection not found")
    
    queries = history.get_history(connection_id, limit)
    return {"success": True, "history": queries, "count": len(queries)}


@router.delete("/connections/{connection_id}/query/history")
async def clear_query_history(connection_id: str):
    """Clear query history for connection."""
    connection = await storage.get_connection(connection_id)
    if not connection:
        raise HTTPException(status_code=404, detail="Connection not found")
    
    success = history.clear_history(connection_id)
    return {"success": success, "message": "History cleared" if success else "No history found"}


@router.get("/connections/{connection_id}/query/history/search")
async def search_query_history(connection_id: str, q: str):
    """Search query history."""
    connection = await storage.get_connection(connection_id)
    if not connection:
        raise HTTPException(status_code=404, detail="Connection not found")
    
    results = history.search_history(connection_id, q)
    return {"success": True, "results": results, "count": len(results)}


@router.post("/query/history/cleanup")
async def cleanup_query_history(retention_days: int = 30):
    """Cleanup old query history based on retention days."""
    removed = history.cleanup_old_history(retention_days)
    return {"success": True, "removed_count": removed, "message": f"Removed {removed} old queries"}


@router.post("/connections/{connection_id}/query/explain")
async def explain_query(connection_id: str, request: QueryRequest):
    """Get query execution plan with AI analysis."""
    from operations.operation_lock import operation_lock

    connection = await storage.get_connection(connection_id)
    if not connection:
        raise HTTPException(status_code=404, detail="Connection not found")

    if operation_lock.is_locked(connection_id):
        raise HTTPException(
            status_code=409, detail="Connection is busy with another operation"
        )

    lock = operation_lock.get_lock(connection_id)
    async with lock:
        # Execute EXPLAIN query
        explain_query_text = f"EXPLAIN (FORMAT JSON, ANALYZE, BUFFERS) {request.query}"
        result = await executor.execute_query(
            connection=connection,
            query=explain_query_text,
            limit=1000,
            offset=0,
            timeout=request.timeout,
        )
        
        if not result["success"]:
            return {"success": False, "error": result.get("error")}
        
        # Get explain output
        explain_output = result["rows"][0][0] if result["rows"] else "{}"
        
        # Analyze with Gemini
        analysis = await analyzer.analyze_explain(request.query, json.dumps(explain_output, indent=2))
        
        return {
            "success": True,
            "explain_plan": explain_output,
            "analysis": analysis.get("analysis") if analysis["success"] else None,
            "error": analysis.get("error") if not analysis["success"] else None
        }

File: src/db-toolkit/electron/backend/handlers/query.js
/**
 * Query execution IPC handlers.
 */

const { ipcMain } = require('electron');
const { connectionStorage } = require('../utils/connection-storage');
const { queryExecutor } = require('../operations/query-executor');
const { queryHistory } = require('../operations/query-history');
const { logger } = require('../utils/logger.js');

function registerQueryHandlers() {
  // Execute query
  ipcMain.handle('query:execute', async (event, connectionId, request) => {
    try {
      const connection = await connectionStorage.getConnection(connectionId);
      if (!connection) {
        throw new Error('Connection not found');
      }

      logger.info(`Executing query on '${connection.name}': ${request.query.substring(0, 100)}...`);

      const result = await queryExecutor.executeQuery(
        connection,
        request.query,
        request.limit,
        request.offset || 0,
        request.timeout,
        request.skipValidation || false
      );

      if (result.success) {
        logger.info(`Query executed successfully (${result.total_rows} rows, ${result.execution_time}s)`);
      } else {
        logger.error(`Query failed: ${result.error}`);
      }

      // Save to history only if not requiring confirmation
      if (!result.requiresConfirmation) {
        await queryHistory.addQuery(
          connectionId,
          request.query,
          result.success,
          result.execution_time,
          result.total_rows,
          result.error
        );
      }

      return { data: result };
    } catch (error) {
      logger.error('Query execution error:', error);
      throw error;
    }
  });

  // Get query history
  ipcMain.handle('query:getHistory', async (event, connectionId, limit = 50) => {
    try {
      const connection = await connectionStorage.getConnection(connectionId);
      if (!connection) {
        throw new Error('Connection not found');
      }

      const queries = await queryHistory.getHistory(connectionId, limit);
      return { data: { success: true, history: queries, count: queries.length } };
    } catch (error) {
      logger.error('Failed to get query history:', error);
      throw error;
    }
  });

  // Clear query history
  ipcMain.handle('query:clearHistory', async (event, connectionId) => {
    try {
      const connection = await connectionStorage.getConnection(connectionId);
      if (!connection) {
        throw new Error('Connection not found');
      }

      const success = await queryHistory.clearHistory(connectionId);
      return {
        data: {
          success,
          message: success ? 'History cleared' : 'No history found',
        }
      };
    } catch (error) {
      logger.error('Failed to clear query history:', error);
      throw error;
    }
  });

  // Delete single query from history
  ipcMain.handle('query:deleteQuery', async (event, connectionId, index) => {
    try {
      const connection = await connectionStorage.getConnection(connectionId);
      if (!connection) {
        throw new Error('Connection not found');
      }

      const success = await queryHistory.deleteQuery(connectionId, index);
      return {
        data: {
          success,
          message: success ? 'Query deleted' : 'Query not found',
        }
      };
    } catch (error) {
      logger.error('Failed to delete query:', error);
      throw error;
    }
  });

  // Search query history
  ipcMain.handle('query:searchHistory', async (event, connectionId, searchTerm) => {
    try {
      const connection = await connectionStorage.getConnection(connectionId);
      if (!connection) {
        throw new Error('Connection not found');
      }

      const results = await queryHistory.searchHistory(connectionId, searchTerm);
      return { data: { success: true, results, count: results.length } };
    } catch (error) {
      logger.error('Failed to search query history:', error);
      throw error;
    }
  });

  // Cleanup old query history
  ipcMain.handle('query:cleanupHistory', async (event, retentionDays = 30) => {
    try {
      logger.info(`Cleaning up query history older than ${retentionDays} days`);
      const removed = await queryHistory.cleanupOldHistory(retentionDays);
      logger.info(`Removed ${removed} old queries from history`);
      return {
        data: {
          success: true,
          removed_count: removed,
          message: `Removed ${removed} old queries`,
        }
      };
    } catch (error) {
      logger.error('Failed to cleanup query history:', error);
      throw error;
    }
  });
}

module.exports = { registerQueryHandlers };
